hashing/src/chunk_with_proof.rs

```diff
@@ -0,0 +1,310 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use casper_types::bytesrepr::{self, Bytes, FromBytes, ToBytes};
+
+use crate::{error::MerkleConstructionError, indexed_merkle_proof::IndexedMerkleProof, Digest};
+
+/// Represents a chunk of data with attached proof.
+#[derive(PartialEq, Debug, JsonSchema, Serialize, Deserialize)]
+#[schemars(with = "String", description = "Chunk of data with attached proof.")]
+#[serde(deny_unknown_fields)]
+pub struct ChunkWithProof {
+    proof: IndexedMerkleProof,
+    #[schemars(with = "String", description = "Hex-encoded bytes.")]
+    chunk: Bytes,
+}
+
+impl ToBytes for ChunkWithProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut result = bytesrepr::allocate_buffer(self)?;
+        result.append(&mut self.proof.to_bytes()?);
+        result.append(&mut self.chunk.to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.proof.serialized_length() + self.chunk.serialized_length()
+    }
+}
+
+impl FromBytes for ChunkWithProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (proof, remainder) = FromBytes::from_bytes(bytes)?;
+        let (chunk, remainder) = FromBytes::from_bytes(remainder)?;
+
+        Ok((ChunkWithProof { proof, chunk }, remainder))
+    }
+}
+
+impl ChunkWithProof {
+    #[cfg(test)]
+    /// 10 bytes for testing purposes.
+    pub const CHUNK_SIZE_BYTES: usize = 10;
+
+    #[cfg(not(test))]
+    /// 1 MiB
+    pub const CHUNK_SIZE_BYTES: usize = 1 << 20;
+
+    /// Constructs the [`ChunkWithProof`] that contains the chunk of data with the appropriate index
+    /// and the cryptographic proof.
+    ///
+    /// Empty data is always represented as single, empty chunk and not as zero chunks.
+    pub fn new(data: &[u8], index: u64) -> Result<Self, MerkleConstructionError> {
+        let (proof, chunk) = match data.is_empty() {
```

Comment:
Just create the `Ok(ChunkWithProof { proof, chunk })` inline. There is no need to destructure a tuple just to then produce this value. Also, this match is unnecessary, and should be rewritten using `if/else`.

Reply:
The destructure-and-construct approach was inherited from the original implementation.
Now fixed, thanks for the good catch.

Regarding the `match` vs `if/else` - I grant the request :)
Rewritten in https://github.com/casper-network/casper-node/pull/2119/commits/735001aed1df3de046222ca9b5602b4153d1bf76